edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Edgelet/ModuleManagementHttpClient.cs

```diff
@@ -3,256 +3,76 @@ namespace Microsoft.Azure.Devices.Edge.Agent.Edgelet
+        internal ModuleManagementHttpClientVersioned GetVersionedModuleManagement(Uri managementUri, string edgeletApiVersion, string edgeletClientApiVersion)
         {
-            try
+            ApiVersion supportedVersion = this.GetSupportedVersion(edgeletApiVersion, edgeletClientApiVersion);
+            if (supportedVersion == ApiVersion.Version20180628)
             {
-                Events.ExecutingOperation(operation, this.managementUri.ToString());
-                T result = await ExecuteWithRetry(func, r => Events.RetryingOperation(operation, this.managementUri.ToString(), r));
-                Events.SuccessfullyExecutedOperation(operation, this.managementUri.ToString());
-                return result;
+                return new Version_2018_06_28.ModuleManagementHttpClient(managementUri);
             }
-            catch (Exception ex)
-            {
-                switch (ex)
-                {
-                    case SwaggerException<ErrorResponse> errorResponseException:
-                        throw new EdgeletCommunicationException($"Error calling {operation}: {errorResponseException.Result?.Message ?? string.Empty}", errorResponseException.StatusCode);
 
-                    case SwaggerException swaggerException:
-                        if (swaggerException.StatusCode < 400)
-                        {
-                            Events.SuccessfullyExecutedOperation(operation, this.managementUri.ToString());
-                            return default(T);
-                        }
-                        else
-                        {
-                            throw new EdgeletCommunicationException($"Error calling {operation}: {swaggerException.Response ?? string.Empty}", swaggerException.StatusCode);
-                        }
-
-                    default:
-                        throw;
-                }
+            if (supportedVersion == ApiVersion.Version20181230)
+            {
+                return new Version_2018_12_30.ModuleManagementHttpClient(managementUri);
             }
-        }
 
-        class ErrorDetectionStrategy : ITransientErrorDetectionStrategy
-        {
-            public bool IsTransient(Exception ex) => ex is SwaggerException se
-                                                     && se.StatusCode >= 500;
+            return new Version_2018_06_28.ModuleManagementHttpClient(managementUri);
         }
 
-        static class Events
+        ApiVersion GetSupportedVersion(string edgeletApiVersion, string edgeletManagementApiVersion)
         {
-            const int IdStart = AgentEventIds.ModuleManagementHttpClient;
-            static readonly ILogger Log = Logger.Factory.CreateLogger<ModuleManagementHttpClient>();
-
-            enum EventIds
-            {
-                ExecutingOperation = IdStart,
-                SuccessfullyExecutedOperation,
-                RetryingOperation
-            }
+            var serverVersion = (ApiVersion)edgeletApiVersion;
```

Comment:
Thanks, I added VersionUnknown, if the client has version Unknown it throws because client should always have a valid version set, if the server version is unknown it will use client version. I think it can be assumed that iotedged will be > edge agent version when server version is unknown, lower version should always be known by edge agent.